### PR TITLE
Mat/prisma

### PR DESCRIPTION
**Pull request by m-mcgowan**
# Problem Context

This ended up being a behemoth of a PR, but I wanted to deliver something that was functional and laid out the foundation for the major changes in the GA release (datastore backend and schema-driven sensor readings.)

## Changes

* PrismaDataStore implementation (wip) displays latest readings
* db-init.ts script that creates a seed database with the built-in sensors
* notehub event webhook that ingests notehub events for sensor readings and populates the database
* changes are done in parallel as much as possible with shared components like ServiceLocator and AppService using conditional logic.

## Testing

Unit / integration / e2e tests?
* e2e tests presently test the beta site only

Steps to test manually?
* Once the development setup is done, browsing `/index2` will show the beginnings of the new site, which displays Nodes with a few additional sensor readings.

## Any other related PRs

#71 #55 

## Ticket(s)

https://www.pivotaltracker.com/n/projects/2534053/stories/181303989

### CODE REVIEW

**Comment by m-mcgowan** [comment]
This is the first pass at integrating Prisma into the project and a first version of the database schema.  We have a shared heroku PostgreSQL instance and instructions for firing up an ephemeral instance locally.

**Comment by m-mcgowan** [comment]
Prisma tries to do some magic regarding generating the ORM model. They recommend you don't check in the generated code, but instead use their preinstall hook which runs `prisma generate`.  As we see, this doesn't work on our docker image. I'm not a huge fan of the "randomness" this introduces, and would much prefer to check in the code as a subfolder for repeatability and consistency.

Thoughts? @paigen11, @tjvantoll, @CarltonHenderson

**Review by CarltonHenderson — COMMENTED** [review]
I don't see where you're running `prisma generate`. But we can fix the docker build I'm sure. Did you add any files the build needs to the (unfortunately named) `.dockerignore` whitelist?

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:9]: Understood. How do we inject the value of the environment variable into the schema? Will that work with `prisma generate`?  I realize it's not ideal, but it doesn't do any harm either.

**Comment by m-mcgowan** [comment]
> 

I didn't make any changes to the whitelist. That could be the problem.

File: DEVELOPMENT.md
Comment (m-mcgowan) [reply DEVELOPMENT.md:37]: I agree, I don't want `dotenv` being used in application code. I was thinking only in the `scripts` section of `package.json` to make `prisma:generate` etc `prisma:migrate` scripts that just work.

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:9]: If we can make it work relatively easily then I'm all for it, but otherwise, it's not worth the effort to change, even if it is slightly architecturally unsound. Looking at `prisma generate` I don't see a way to specify the URL on the command line, which would have been better. `PrismaClient` does allow the db url to be specified.

The advantage of doing it the way it is now is that it follows the pattern laid out in all their examples, so anyone onboarding will immediately recognize what we are doing.

Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:9]: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#prismaclient.  btw, the `PrismaClient` will be injected into the `PrismaDataProvider` instance (or at the very least, the datasource URL.). 

That still leaves us with the default url in the schema which is pulled from the environment, but that's how these tools work - changing that looks to be a big lift and not worth the payback imho.

File: DEVELOPMENT.md
Comment (paigen11) [inline DEVELOPMENT.md:56]: Include link to Notion docs here?

File: DEVELOPMENT.md
Comment (m-mcgowan) [reply DEVELOPMENT.md:43]: Even if it doesn't work?! 👿  😜

File: yarn.lock
Comment (paigen11) [inline yarn.lock:383]: Did we decide to no longer include lock files in individual PR branches?

Comment (m-mcgowan) [reply yarn.lock:383]: We didn't yet. We had the discussion but no clear conclusion. It would be my preference to leave them out - they are a deployment/DevOps/security concern not so much day to day development.  But seems most people on the net haven't realized that yet ;-)

File: DEVELOPMENT.md
Comment (paigen11) [reply DEVELOPMENT.md:43]: Oh, well if that's the case, forget it :)

**Comment by paigen11** [comment]
> Prisma tries to do some magic regarding generating the ORM model. They recommend you don't check in the generated code, but instead use their preinstall hook which runs `prisma generate`. As we see, this doesn't work on our docker image. I'm not a huge fan of the "randomness" this introduces, and would much prefer to check in the code as a subfolder for repeatability and consistency.
> 
> Thoughts? @paigen11, @tjvantoll, @CarltonHenderson

If that's Prisma's preferred way that we use it, I'd like to defer to them and believe that if we lock down the version of Prisma we're using in our `package.json` and `yarn.lock` we won't run into gotchas as a result of breaking changes in future Prisma releases, but I can be persuaded otherwise.

File: DEVELOPMENT.md
Comment (paigen11) [reply DEVELOPMENT.md:56]: Got it. Thanks for clarifying

File: .dockerignore
Comment (CarltonHenderson) [inline .dockerignore:8]:
```suggestion
!node_modules/.prisma/
```

Comment (CarltonHenderson) [reply .dockerignore:8]: Actually, this doesn't make sense either way... node_modules/.prisma isn't committed to git so there's no effect from whitelisting `!...` or blacklisting it.

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:16]: Company name, building? Something more meaningful than the notehub ProjectUID.

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:35]: +1

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:45]: Ah, my memory failed.  Should be `deviceEUI`. What we've been calling the mac address is the LoRa device EUI.

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:50]: description is longer and only shown on the details page. Label is meant as a short tag for the device. Like a subheading or additional short info that is shown in summary, or via tooltip etc.

We've been focusing on nodes and gateways, but really our customers will be focusing on the kind of products they build. A label can be used to supplement with additional product info (e.g. the type of device.)

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:68]: A human readable description for the type of sensor generating the data. For example, a project may have 5 sensors each reading the value of a different analog input. The description helps disambiguate and clarify  what values are being read. The longer description can be useful for installers, support personnel etc.

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:76]: yes :).  I am still uncertain of the best representation and will iterate on this.  I'll probably start with a simple number, but I know that's not our final stop.

File: package.json
Comment (m-mcgowan) [reply package.json:16]: it wasn't working at one point. I can revert.

File: .dockerignore
Comment (m-mcgowan) [reply .dockerignore:8]: I'll remove it and see if everything still works.  I added this at the point when the docker build was broken and we didn't know why.  It should all just work without the explicit call to `prisma generate`, but that's not the case.

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:71]: It's just an idea I'm toying with. If we don't need it I can remove it from the schema.  When I first created the schema they were not tied to a specific device, but now they are, so description makes a little less sense.

File: src/services/prisma-datastore/schema.prisma
Comment (tjvantoll) [reply src/services/prisma-datastore/schema.prisma:16]: Gotcha. Ok, that makes sense 👍

File: src/services/prisma-datastore/schema.prisma
Comment (tjvantoll) [reply src/services/prisma-datastore/schema.prisma]: Got it 👍

File: src/services/prisma-datastore/schema.prisma
Comment (tjvantoll) [reply src/services/prisma-datastore/schema.prisma:71]: Ah, yeah I could see that in the future. Personally I prefer we hold off on things like this until we have a need for it in the app, but not a huge deal either way.

File: src/services/prisma-datastore/db-init.ts
Comment (CarltonHenderson) [inline src/services/prisma-datastore/db-init.ts:3]: It'd be better to have the Config depend on Prisma, not the other way around. Otherwise this depends on Config which depends on everything and then everything's dependent on everything. The config is the lowest level implementation detail in other words.

For Example, imagine in the future if we have some sort of 'context' defining the project, company name, etc. it would suck to have to rewrite this component.

File: src/services/prisma-datastore/schema.prisma
Comment (CarltonHenderson) [reply src/services/prisma-datastore/schema.prisma:9]: I think it'd be better to override it in code when creating the PrismaClient 

https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#programmatically-override-a-datasource-url

File: src/services/prisma-datastore/schema.prisma
Comment (CarltonHenderson) [inline src/services/prisma-datastore/schema.prisma:14]: Not a fan of tracking things that aren't in service of a user story. What's the justification for `createdAd`?

File: src/services/prisma-datastore/schema.prisma
Comment (CarltonHenderson) [inline src/services/prisma-datastore/schema.prisma:47]:
```suggestion
  nodeID      String       @unique @map("node_id") @db.Char(64)
```

File: src/services/prisma-datastore/schema.prisma
Comment (CarltonHenderson) [inline src/services/prisma-datastore/schema.prisma:53]:
```suggestion
  lastSeenAt      DateTime?   @map("last_seen_at") // the last time an event was heard from this device
```

I guess even if it's just a session, it'll leave a new event in _session.qo. Are we sure that's the right semantic meaning though? Maybe I'm splitting hairs.

File: src/services/prisma-datastore/schema.prisma
Comment (CarltonHenderson) [inline src/services/prisma-datastore/schema.prisma:58]:
```suggestion
  readingSchemas  ReadingSchema[]
```

Are we going to need `Sensors` in the database?

File: src/services/prisma-datastore/schema.prisma
Comment (CarltonHenderson) [inline src/services/prisma-datastore/schema.prisma:70]:
```suggestion
    notefile     String              @db.VarChar(32) @map("notefile")  // "air.qo"
```

or

```suggestion
    eventQueueNotefile     String              @db.VarChar(32) @map("event_queue_notefile")  // "air.qo"
```

something else? eventName doesn't ring well with me.

File: src/services/prisma-datastore/schema.prisma
Comment (CarltonHenderson) [inline src/services/prisma-datastore/schema.prisma:72]:
```suggestion
    presentationUnit          String?             @db.Char(10) // e.g. kPa
    storageUnit          String?             @db.Char(10) // e.g. Pa
```

File: src/services/prisma-datastore/schema.prisma
Comment (CarltonHenderson) [inline src/services/prisma-datastore/schema.prisma:85]:
```suggestion
    schema        ReadingSchema   @relation(fields: [schema_id], references: [id])
```
`Device` doesn't add any value in my mind.

**Review by CarltonHenderson — COMMENTED** [review]
comments attached

File: src/services/prisma-datastore/db-init.ts
Comment (m-mcgowan) [reply src/services/prisma-datastore/db-init.ts:3]: I see what you mean, but not how it plays out. Presently, Config just depends on the `process.env`, which is how we are abstracting the external environment. Perhaps it should be renamed `Environment` and config is something else that we don't yet have (probably best embodied by the bits of code in ServiceLocator that builds the AppService. )

**Comment by m-mcgowan** [comment]
image

WiP on dynamic sensor types.  Still some tidy up to do, but this should be a good place for everyone to start biting off a piece :-)

File: __tests__/src/services/notehub/NotehubEvents.test.ts
Comment (paigen11) [inline __tests__/src/services/notehub/NotehubEvents.test.ts:21]:
```suggestion
            it("returns the event name unmodifed", () => {
```

File: cypress/cypress/integration/homepage.spec.js
Comment (paigen11) [inline cypress/cypress/integration/homepage.spec.js:1]: this nested folder of `cypress/cypress/` should not exist. Delete please

File: src/components/elements/FlexibleGatewayCard.tsx
Comment (paigen11) [inline src/components/elements/FlexibleGatewayCard.tsx:21]: This should probably be moved either into the body of this component or into a separate helper file and then called here, based on if other components will need to access it too

File: src/components/elements/FlexibleNodeCard.tsx
Comment (paigen11) [inline src/components/elements/FlexibleNodeCard.tsx:3]: move scss imports to the bottom

File: src/components/elements/FlexibleNodeCard.tsx
Comment (paigen11) [inline src/components/elements/FlexibleNodeCard.tsx:38]: remove console.log?

File: src/components/elements/FlexibleNodeCard.tsx
Comment (paigen11) [inline src/components/elements/FlexibleNodeCard.tsx:55]: remove console.logs

File: src/components/elements/FlexibleNodeCard.tsx
Comment (paigen11) [inline src/components/elements/FlexibleNodeCard.tsx:49]: remove console.log

File: src/components/elements/NoopReadingRenderer.tsx
Comment (paigen11) [inline src/components/elements/NoopReadingRenderer.tsx:2]: what's the purpose of this component?

File: src/services/prisma-datastore/schema.prisma
Comment (m-mcgowan) [reply src/services/prisma-datastore/schema.prisma:8]: This is done. https://github.com/blues/sparrow-reference-web-app/blob/17535c0d521fc8e1f423e49bc31707312ba6dc50/src/services/ServiceLocator.ts#L67

File: src/components/elements/FlexibleNodeCard.tsx
Comment (m-mcgowan) [reply src/components/elements/FlexibleNodeCard.tsx:38]: This is still WiP but when we get closer to the code being production ready we'll remove it. (Along with cleaning up other lint complaints!)

File: src/components/elements/NoopReadingRenderer.tsx
Comment (m-mcgowan) [reply src/components/elements/NoopReadingRenderer.tsx:2]: It's for when you don't want to render a particular type of sensor. It was a last minute change to try to not render the count total as a separate tile, but it didn't work as expected. We can iterate on this.  It may be better handled in a different way.

File: src/components/models/Gateway.ts
Comment (paigen11) [inline src/components/models/Gateway.ts:18]: does a gateway need to know about sensor readings? Or just the sensor model itself?

File: src/services/DomainModel.ts
Comment (paigen11) [inline src/services/DomainModel.ts:315]: what does RSSI stand for? It's not an acronym I'm familiar with

File: src/services/AppModelBuilder.ts
Comment (paigen11) [inline src/services/AppModelBuilder.ts:135]:
```suggestion
        return visitor.visitGateway({
```

File: cypress/cypress/integration/homepage.spec.js
Comment (m-mcgowan) [reply cypress/cypress/integration/homepage.spec.js:1]: I had meant to do this - I think it was from choosing the wrong side of a merge.

File: DEVELOPMENT.md
Comment (tjvantoll) [inline DEVELOPMENT.md:2]: Presumably as that gets merged this step will become required? If so we’ll want to list this as a required setup step in our README.

Also I can make a pass through this doc after it’s merged. Don’t think it’s worth getting into small things like formatting in this PR.

File: src/components/models/Gateway.ts
Comment (m-mcgowan) [reply src/components/models/Gateway.ts:18]: This is the old model - it will be phased out eventually. I'll remove the comment. The intent is that Gateways have other "sensors", like wifi/cellular signal strength. These are present in the new AppModel.

File: src/components/elements/FlexibleGatewayCard.tsx
Comment (tjvantoll) [inline src/components/elements/FlexibleGatewayCard.tsx:40]: The build fails on this line with this error.

> Type error: Property 'numericValue' does not exist on type 'Reading'.

File: src/services/DomainModel.ts
Comment (m-mcgowan) [reply src/services/DomainModel.ts:315]: Received Signal Strength Indicator - it's a measure of signal strength.

File: src/services/prisma-datastore/PrismaDataProvider.ts
Comment (tjvantoll) [inline src/services/prisma-datastore/PrismaDataProvider.ts:1]: For me locally TypeScript doesn’t like any of these imports. For example.

> Module '"@prisma/client"' has no exported member 'PrismaClient'.

Comment (m-mcgowan) [reply src/services/prisma-datastore/PrismaDataProvider.ts:1]: did you do `yarn install`? That should create what is needed.

**Review by CarltonHenderson — APPROVED** [review]
Approved w? some comments I wrote up before the review meeting. FInally figured out how to submit them I think!

File: src/services/prisma-datastore/PrismaDataProvider.ts
Comment (tjvantoll) [reply src/services/prisma-datastore/PrismaDataProvider.ts:1]: fyi the errors did go away after I restarted Visual Studio Code. Something to keep in mind if others hit this same problem.

File: src/components/elements/FlexibleGatewayCard.tsx
Comment (m-mcgowan) [reply src/components/elements/FlexibleGatewayCard.tsx]: Let's iterate on it after this PR is merged.

File: src/pages/index2.tsx
Comment (m-mcgowan) [reply src/pages/index2.tsx]: Renamed it to dev.tsx...it's our work in progress development site, while index.tsx is more stable.

File: src/components/elements/FlexibleGatewayCard.tsx
Comment (m-mcgowan) [reply src/components/elements/FlexibleGatewayCard.tsx]: Fixed...it was a last minute change that I hadn't done completely.

File: src/components/renderers/builtin.ts
Comment (m-mcgowan) [reply src/components/renderers/builtin.ts]: I've reworked this a little to hopefully make it clearer.

File: src/components/renderers/renderers.ts
Comment (m-mcgowan) [reply src/components/renderers/renderers.ts]: Consider it an optimization - the system will attempt to display the data but it may not be optimal. Developers can choose to add their own renderer if they are not happy with the default output.

File: src/components/renderers/renderers.ts
Comment (m-mcgowan) [reply src/components/renderers/renderers.ts]: Sensors also appear on Gateways (e.g. signal strength, voltage, temperature). I wanted to be able to provide as much context as possible to the renderer.  SensorHost is ok, but it's abstract and there's no easy way to get back to the actual Gateway/Node or know the Gateway a Node belongs to. 

The intent is to provide a rich model the renderer can use beyond just looking at the reading or reading series it has to render.

Comment (m-mcgowan) [reply src/components/renderers/renderers.ts]: Target may be a poor naming choice. We can rename when we find a better one. :-)

**Pull request merged**
